server/asset/btc/addresser.go

```diff
@@ -0,0 +1,85 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+package btc
+
+import (
+	"errors"
+	"fmt"
+
+	"decred.org/dcrdex/server/asset"
+	"github.com/btcsuite/btcd/chaincfg"
+	"github.com/btcsuite/btcutil"
+	"github.com/btcsuite/btcutil/hdkeychain"
+)
+
+// AddressDeriver generates unique addresses from an extended public key.
+type AddressDeriver struct {
+	params   *chaincfg.Params
+	storeIdx func(uint32) // e.g. callback to a DB update
+	xpub     *hdkeychain.ExtendedKey
+	next     uint32 // next child index to generate
+}
+
+// NewAddressDeriver creates a new NewAddressDeriver AddressDeriver for the
+// provided extended public key, HDKeyIndexer, and network parameters.
+func NewAddressDeriver(xpub string, keyIndexer asset.HDKeyIndexer, chainParams *chaincfg.Params) (*AddressDeriver, error) {
+	key, err := hdkeychain.NewKeyFromString(xpub)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing master pubkey: %w", err)
+	}
+	if !key.IsForNet(chainParams) {
+		return nil, fmt.Errorf("key is for the wrong network: %s", chainParams.Name)
+	}
+	if key.IsPrivate() {
+		return nil, errors.New("private key provided")
+	}
+	external, err := key.Child(0)
+	if err != nil {
+		return nil, err
+	}
+	pubkey, _ := external.ECPubKey()
+	pkb := pubkey.SerializeCompressed()
```

Comment:
To anyone reviewing just this commit rather than the entire PR, this was changed in the "FIX" commit to go back to indexing on the xpub string.